numpy/core/src/multiarray/arrayobject.c

```diff
@@ -500,10 +501,11 @@ array_dealloc(PyArrayObject *self)
             char const * msg = "UPDATEIFCOPY detected in array_dealloc. "
                 " Required call to PyArray_ResolveWritebackIfCopy or "
                 "PyArray_DiscardWritebackIfCopy is missing";
-            Py_INCREF(self); /* hold on to self in next call  since if
-                              * refcount == 0 it will recurse back into
-                              *array_dealloc
-                              */
+            /*
+             * prevent reaching 0 twice and thus recursing into dealloc.
+             * Avoid INCREF for now to avoid increasing sys.gettotalrefcount.
+             */
+            PyArray_REFCOUNT(self) += 1;
```

Comment:
This is a necessary evil to avoid changing the total reference count, and it should work to prevent the array reaching a reference count of 0 twice. I do not like the hidden struct access, but we have it in a different code path already, so my guess it is just as well (until it breaks).

Reply:
Ah, cannot use that macro on PyPy, but I guess the strcut access probably works...

Reply:
That macro is an evil hack. Python is moving away from direct access, this will break if CPython ever moves forward with its garbage collecting. We should remove it. Can't you just add a DECREF later?

Reply:
As discussed, I do not know. The first two paths we can maybe avoid easily, the last one maybe an ifdef can be done, or we just wait until python actually breaks us and fix it up then. In Principle once the refcount reaches 0, it should probably never increase again, but we need to iterate the array and that uses so much machinery which holds on to references, that it might be annoying to fix for real.

Reply:
Can we split out these scary-looking `PyArray_REFCOUNT` changes from the obviously correct ones?

Reply:
Will split it off. As a random remark: I just saw that cython does the same thing (i.e. modify reference count without using `INCREF`.